// @langchain/core/dist/utils/uuid/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_runtime = require("../../_virtual/_rolldown/runtime.cjs");
const require_max = require("./max.cjs");
const require_nil = require("./nil.cjs");
const require_validate = require("./validate.cjs");
const require_parse = require("./parse.cjs");
const require_stringify = require("./stringify.cjs");
const require_v1 = require("./v1.cjs");
const require_v4 = require("./v4.cjs");
const require_v5 = require("./v5.cjs");
const require_v6 = require("./v6.cjs");
const require_v7 = require("./v7.cjs");
const require_version = require("./version.cjs");
//#region src/utils/uuid/index.ts
var uuid_exports = /* @__PURE__ */ require_runtime.__exportAll({
	MAX: () => MAX,
	NIL: () => NIL,
	parse: () => parse,
	stringify: () => stringify,
	v1: () => v1,
	v4: () => v4,
	v5: () => v5,
	v6: () => v6,
	v7: () => v7,
	validate: () => validate,
	version: () => version
});
const MAX = require_max.default;
const NIL = require_nil.default;
const parse = require_parse.default;
const stringify = require_stringify.default;
const v1 = require_v1.default;
const v4 = require_v4.default;
const v5 = require_v5.default;
const v6 = require_v6.default;
const v7 = require_v7.default;
const validate = require_validate.default;
const version = require_version.default;
//#endregion
exports.MAX = MAX;
exports.NIL = NIL;
exports.parse = parse;
exports.stringify = stringify;
Object.defineProperty(exports, "uuid_exports", {
	enumerable: true,
	get: function() {
		return uuid_exports;
	}
});
exports.v1 = v1;
exports.v4 = v4;
exports.v5 = v5;
exports.v6 = v6;
exports.v7 = v7;
exports.validate = validate;
exports.version = version;
