// @langchain/core/dist/utils/uuid/index.js
import { __exportAll } from "../../_virtual/_rolldown/runtime.js";
import max_default from "./max.js";
import nil_default from "./nil.js";
import validate$1 from "./validate.js";
import parse$1 from "./parse.js";
import stringify$1 from "./stringify.js";
import v1$1 from "./v1.js";
import v4$1 from "./v4.js";
import v5$1 from "./v5.js";
import v6$1 from "./v6.js";
import v7$1 from "./v7.js";
import version$1 from "./version.js";
//#region src/utils/uuid/index.ts
var uuid_exports = /* @__PURE__ */ __exportAll({
	MAX: () => MAX,
	NIL: () => NIL,
	parse: () => parse,
	stringify: () => stringify,
	v1: () => v1,
	v4: () => v4,
	v5: () => v5,
	v6: () => v6,
	v7: () => v7,
	validate: () => validate,
	version: () => version
});
const MAX = max_default;
const NIL = nil_default;
const parse = parse$1;
const stringify = stringify$1;
const v1 = v1$1;
const v4 = v4$1;
const v5 = v5$1;
const v6 = v6$1;
const v7 = v7$1;
const validate = validate$1;
const version = version$1;
//#endregion
export { MAX, NIL, parse, stringify, uuid_exports, v1, v4, v5, v6, v7, validate, version };
