// @langchain/langgraph-sdk/dist/stream/optimistic-input.js
import { ensureMessageInstances } from "./message-coercion.js";
import { toMessageDict } from "../ui/messages.js";
//#region src/stream/optimistic-input.ts
function isBaseMessageInstance(value) {
	return value != null && typeof value.getType === "function";
}
function extractId(value) {
	const id = value?.id;
	return typeof id === "string" && id.length > 0 ? id : void 0;
}
/**
* Normalize a message-key value into an array of entries. Mirrors the
* server's `add_messages` coercion: a bare string or single message
* object is treated as a one-element list.
*/
function toEntryArray(value) {
	if (Array.isArray(value)) return value;
	return [value];
}
/**
* Build a message dict carrying `id` from an arbitrary input entry,
* without mutating the original.
*/
function toDispatchDict(entry, id) {
	if (typeof entry === "string") return {
		type: "human",
		content: entry,
		id
	};
	if (isBaseMessageInstance(entry)) return {
		...toMessageDict(entry),
		id
	};
	return {
		...entry,
		id
	};
}
/**
* Prepare a raw submit input for optimistic dispatch.
*
* @param raw         - Raw input passed to `submit()`. Must be a
*   non-null, non-array object (caller guards this).
* @param messagesKey - State key holding the message array.
* @param mintId      - Factory for stable client message ids.
* @returns The dispatch payload, optimistic messages, echoed ids, and
*   the non-message portion of the input.
*/
function prepareOptimisticInput(raw, messagesKey, mintId) {
	const dispatchInput = { ...raw };
	const extraValues = {};
	for (const key of Object.keys(raw)) if (key !== messagesKey) extraValues[key] = raw[key];
	const echoedIds = [];
	const messagesValue = raw[messagesKey];
	if (messagesValue == null) return {
		dispatchInput,
		optimisticMessages: [],
		echoedIds,
		extraValues
	};
	const entries = toEntryArray(messagesValue);
	const dispatchEntries = [];
	const optimisticDicts = [];
	for (const entry of entries) {
		if (!(typeof entry === "string" || isBaseMessageInstance(entry) || entry != null && typeof entry === "object" && !Array.isArray(entry))) {
			dispatchEntries.push(entry);
			continue;
		}
		const id = extractId(entry) ?? mintId();
		const dict = toDispatchDict(entry, id);
		dispatchEntries.push(dict);
		optimisticDicts.push(dict);
		echoedIds.push(id);
	}
	dispatchInput[messagesKey] = dispatchEntries;
	return {
		dispatchInput,
		optimisticMessages: ensureMessageInstances(optimisticDicts),
		echoedIds,
		extraValues
	};
}
//#endregion
export { prepareOptimisticInput };
